tests/integration/jupyter_tests.rs

```diff
@@ -0,0 +1,530 @@
+      "--conn",
+      conn_file.to_string().as_str(),
+    ])
+    .spawn()
+    .unwrap();
+  (context, conn, JupyterServerProcess(Some(process)))
+}
+
+async fn setup() -> (TestContext, JupyterClient, JupyterServerProcess) {
+  let (context, conn, process) = setup_server();
+  let client = JupyterClient::new(&conn).await;
+  client.io_subscribe("").await.unwrap();
+
+  (context, client, process)
+}
+
+/// Asserts that the actual value is equal to the expected value, but
+/// only for the keys present in the expected value.
+/// In other words, `assert_eq_subset(json!({"a": 1, "b": 2}), json!({"a": 1}))` would pass.
+#[track_caller]
+fn assert_eq_subset(actual: Value, expected: Value) {
+  match (actual, expected) {
+    (Value::Object(actual), Value::Object(expected)) => {
+      for (k, v) in expected.iter() {
+        let Some(actual_v) = actual.get(k) else {
+          panic!("Key {k:?} not found in actual value ({actual:#?})");
+        };
+        assert_eq_subset(actual_v.clone(), v.clone());
+      }
+    }
+    (Value::Array(actual), Value::Array(expected)) => {
+      for (i, v) in expected.iter().enumerate() {
+        assert_eq_subset(actual[i].clone(), v.clone());
+      }
+    }
+    (actual, expected) => {
+      assert_eq!(actual, expected);
+    }
+  }
+}
+
+#[tokio::test]
+async fn jupyter_heartbeat_echoes() -> Result<()> {
+  let (_ctx, client, _process) = setup().await;
+  client.send_heartbeat(b"ping").await?;
+  let msg = client.recv_heartbeat().await?;
+  assert_eq!(msg, Bytes::from_static(b"ping"));
+
+  Ok(())
+}
+
+#[tokio::test]
+async fn jupyter_kernel_info() -> Result<()> {
+  let (_ctx, client, _process) = setup().await;
+  client
+    .send(Control, "kernel_info_request", json!({}))
+    .await?;
+  let msg = client.recv(Control).await?;
+  assert_eq!(msg.header.msg_type, "kernel_info_reply");
+  assert_eq_subset(
+    msg.content,
+    json!({
+      "status": "ok",
+      "implementation": "Deno kernel",
+      "language_info": {
+        "name": "typescript",
+        "mimetype": "text/x.typescript",
+        "file_extension": ".ts",
+        "pygments_lexer": "typescript",
+        "nb_converter": "script"
+      },
+    }),
+  );
+
+  Ok(())
+}
+
+#[tokio::test]
+async fn jupyter_execute_request() -> Result<()> {
+  let (_ctx, client, _process) = setup().await;
+  let request = client
+    .send(
+      Shell,
+      "execute_request",
+      json!({
+        "silent": false,
+        "store_history": true,
+        "user_expressions": {},
+        "allow_stdin": true,
+        "stop_on_error": false,
+        "code": "console.log(\"asdf\")"
+      }),
+    )
+    .await?;
+  let reply = client.recv(Shell).await?;
+  assert_eq!(reply.header.msg_type, "execute_reply");
+  assert_eq_subset(
+    reply.content,
+    json!({
+      "status": "ok",
+      "execution_count": 1,
+    }),
+  );
+
+  let execution_busy = client.recv(IoPub).await?;
+  assert_eq!(execution_busy.header.msg_type, "status");
```

Comment:
I think this might be racy - since you're using `console.log` in the code you'll have both "status" and "stream" messages. Maybe capture two messages from IoPub and assert that you got one of each in no particular order?

Reply:
Yup definitely, should be fixed now.